pkg/controller/prune.go

```diff
@@ -0,0 +1,65 @@
+package controller
+
+import (
+	"fmt"
+
+	n "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2018-12-01/network"
+	"github.com/golang/glog"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+
+	"github.com/Azure/application-gateway-kubernetes-ingress/pkg/annotations"
+	"github.com/Azure/application-gateway-kubernetes-ingress/pkg/appgw"
+	"github.com/Azure/application-gateway-kubernetes-ingress/pkg/brownfield"
+	"github.com/Azure/application-gateway-kubernetes-ingress/pkg/events"
+)
+
+type pruneFunc func(appGw *n.ApplicationGateway, cbCtx *appgw.ConfigBuilderContext, ingressList []*v1beta1.Ingress) []*v1beta1.Ingress
+
+// PruneIngress filters ingress list based on filter functions and returns a filtered ingress list
+func (c AppGwIngressController) PruneIngress(appGw *n.ApplicationGateway, cbCtx *appgw.ConfigBuilderContext) []*v1beta1.Ingress {
+
+	pruneFuncList := []pruneFunc{
+		c.pruneNoPrivateIP,
+	}
+
+	if cbCtx.EnvVariables.EnableBrownfieldDeployment == "true" {
+		pruneFuncList = append(pruneFuncList, c.pruneProhibitedIngress)
+	}
+
+	prunedIngressListed := cbCtx.IngressList
+	for _, prune := range pruneFuncList {
+		prunedIngressListed = prune(appGw, cbCtx, prunedIngressListed)
+	}
+
+	return prunedIngressListed
+}
+
+func (c AppGwIngressController) pruneProhibitedIngress(appGw *n.ApplicationGateway, cbCtx *appgw.ConfigBuilderContext, ingressList []*v1beta1.Ingress) []*v1beta1.Ingress {
+	// Mutate the list of Ingresses by removing ones that AGIC should not be creating configuration.
+	for idx, ingress := range ingressList {
+		glog.V(5).Infof("Original Ingress[%d] Rules: %+v", idx, ingress.Spec.Rules)
+		ingressList[idx].Spec.Rules = brownfield.PruneIngressRules(ingress, cbCtx.ProhibitedTargets)
+		glog.V(5).Infof("Sanitized Ingress[%d] Rules: %+v", idx, ingress.Spec.Rules)
+	}
+
+	return ingressList
+}
+
+func (c AppGwIngressController) pruneNoPrivateIP(appGw *n.ApplicationGateway, cbCtx *appgw.ConfigBuilderContext, ingressList []*v1beta1.Ingress) []*v1beta1.Ingress {
+	prunedIngressList := make([]*v1beta1.Ingress, 0)
+	appGwHasPrivateIP := appgw.GetIPConfigurationID(appGw, true) != nil
+	for _, ingress := range ingressList {
+		usePrivateIP, _ := annotations.UsePrivateIP(ingress)
+		usePrivateIP = usePrivateIP || cbCtx.EnvVariables.UsePrivateIP == "true"
+		if usePrivateIP && !appGwHasPrivateIP {
+			errorLine := fmt.Sprintf("Removing Ingress %s/%s with '%s: true' annotation as Application Gateway doesn't have a private IP", ingress.Namespace, ingress.Name, annotations.UsePrivateIPKey)
```

Comment:
It would be great to make this error message actionable. I'd **add** something like `Add a private IP address to App Gateway %s to use it with ingress %s`
It might be helpful to say `Ingress %s requires App Gwy %s has a private IP adress` as well...

It would be great if this log message answers the question: "How do I fix this?"